### DISCUSSION TITLE
[Kanban] List history for cards

### DISCUSSION THREAD

**Discussion by AshRubigo** [BurningTreeC/tiddlywiki-muuri#19] [Ideas]
This might be a more difficult idea.

Currently cards don't track *when* they were added to a particular column. This is useful in determining how long a task took, or whether it was completed before or after another, as examples.

A solution might be to have a 'kanban column history' field for all the cards, holding items that refer to the date of the move and the title of the column it was moved to, such as `[[YYYY0MM0DD0hh0mm0ss0XXX||Column tiddler title]]`. The contents of the field could be revealed by pressing a button on each card, showing something like:

```
History:
(2021/01/02 01:01:01:001) Column tiddler title
(2021/01/01 01:01:01:001) A different list tiddler
```

**Comment by BurningTreeC** [comment]
Thanks for the idea @AshRubigo

I'll have to think about this a bit.

Would you mind giving me some examples how this could be used?

**Reply by AshRubigo** [reply]
Kanban boards are great for displaying ***what***, but not so much ***when***. Ultimately the purpose is to help in recalling the state of some project, and 'when' can be just as useful picking up where you left off after weeks/months away. I might ask:

* Which of these 2 completed tasks did I start doing first?
* Which of these 2 completed tasks took the longest?
* Was this task previously complete, but returned to in-progress, and when did that happen?
* Was this card always low priority?
* Did I start working on these 2 tasks at a similar time?
* Did I ever start that task, or did I move it back to the 'todo' column at some point?

Further, even if you didn't want to add any additional time-related features to the plugin, since a fully fledged project manager likely wasn't your initial plan, history tracking would be very nifty for those extending the plugin with custom functionality. I'd like to have a changelog of sorts for all my kanban boards, and a more extravagant version of that would be a gantt chart.

**Reply by BurningTreeC** [reply]
Ok thanks for the explanation :)

- a newly added item should get a timestamp and a reference to the column it was added to
- when moving items between columns the items should also get a timestamp and a reference like `from-to`

did I get this right?

**Reply by AshRubigo** [reply]
Not even `from-to`. Let's take this example of a `kanban column history` field of a card (note `[[||]]` might not be the best format):

```
[[20210102000000000||Column B]] [[20210101000000000||Column A]]
```

Here we have a history of the columns a card was in. Each entry in this list has the **date** and **location** of the move, and that is all that needs to be recorded. From this we know at midnight 1st January 2021 the card was created in `Column A`, then exactly 24 hours later it was moved to `Column B`. By extension we know that the card was moved from `Column A` to `Column B`, and when that happened.

That information can then be used in any way by you or the user, such as reformatted to be more human-readable in a change log for a board with a macro `<<kanban-board-history>>`:

```
!! History

* 2021/01/02 12:00:00:000 - [[Card B]] to [[Column C]]
* 2021/01/02 00:00:00:000 - [[Card A]] to [[Column B]]
* 2021/01/01 08:00:00:000 - [[Card B]] to [[Column A]]
* 2021/01/01 00:00:00:000 - [[Card A]] to [[Column A]]
```

**Comment by BurningTreeC** [comment]
Hi @AshRubigo, I've now added this functionality but the `<<kanban-board-history>>` macro is missing

Whenever a card gets created or added to a column the `kanban-history` field of a card gets populated with a string of the form:

```
[[20210102-120000000,column: columnName]]
```

**Reply by AshRubigo** [reply]
Wow, that's amazing turnaround!

I would recommend removing the hyphen in the middle of the date, since then it would be in-line with the format of the 'created' field of tiddlers.

**Reply by BurningTreeC** [reply]
Thanks @AshRubigo , I've removed the hyphen now. It might be that this works only with the current prerelease of tiddlywiki. I haven't tested it with the current 5.1.23 release
